lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,747 @@
+
+export type TypeInputGetOrigin = (input: {
+    tenantId: string;
+    request: BaseRequest;
+    userContext: UserContext;
+}) => Promise<string>;
+
+export type TypeInputValidateEmailAddress = (
+    email: string,
+    tenantId: string
+) => Promise<string | undefined> | string | undefined;
+
+// centralize error types in order to prevent missing cascading errors
+// type RegisterOptionsErrorResponse = { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" };
+
+// type SignInOptionsErrorResponse = { status: "WRONG_CREDENTIALS_ERROR" };
+
+// type SignUpErrorResponse =
+//     | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+//     | { status: "WRONG_CREDENTIALS_ERROR" }
+//     | { status: "INVALID_AUTHENTICATOR_ERROR"; reason: string };
+
+// type SignInErrorResponse = { status: "WRONG_CREDENTIALS_ERROR" };
+
+// type VerifyCredentialsErrorResponse = { status: "WRONG_CREDENTIALS_ERROR" };
+
+// type GenerateRecoverAccountTokenErrorResponse = { status: "UNKNOWN_USER_ID_ERROR" };
+
+// type ConsumeRecoverAccountTokenErrorResponse = { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" };
+
+// type RegisterCredentialErrorResponse =
+//     | { status: "WRONG_CREDENTIALS_ERROR" }
+//     // when the attestation is checked and is not valid or other cases in whcih the authenticator is not correct
+//     | { status: "INVALID_AUTHENTICATOR_ERROR"; reason: string };
+
+// type CreateNewRecipeUserErrorResponse =
+//     | { status: "WRONG_CREDENTIALS_ERROR" }
+//     | { status: "INVALID_AUTHENTICATOR_ERROR"; reason: string }
+//     | { status: "EMAIL_ALREADY_EXISTS_ERROR" };
+
+// type DecodeCredentialErrorResponse = { status: "WRONG_CREDENTIALS_ERROR" };
+
+// type GetUserFromRecoverAccountTokenErrorResponse = { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" };
+
+// type RemoveCredentialErrorResponse = { status: "CREDENTIAL_NOT_FOUND_ERROR" };
+
+// type GetCredentialErrorResponse = { status: "CREDENTIAL_NOT_FOUND_ERROR" };
+
+// type RemoveGeneratedOptionsErrorResponse = { status: "GENERATED_OPTIONS_NOT_FOUND_ERROR" };
+
+// type GetGeneratedOptionsErrorResponse = { status: "GENERATED_OPTIONS_NOT_FOUND_ERROR" };
+
+type Base64URLString = string;
+
+export type RecipeInterface = {
+    // should have a way to access the user email: passed as a param, through session, or using recoverAccountToken
+    // it should have at least one of those 3 options
+    registerOptions(
+        input: {
+            relyingPartyId: string;
+            relyingPartyName: string;
+            origin: string;
+            requireResidentKey: boolean | undefined; // should default to false in order to allow multiple authenticators to be used; see https://auth0.com/blog/a-look-at-webauthn-resident-credentials/
+            // default to 'required' in order store the private key locally on the device and not on the server
+            residentKey: "required" | "preferred" | "discouraged" | undefined;
```

Comment:
Since these lists seem to be reused in a few places (e.g.: both in the input and the return types) I'd prefer if you moved made these into re-usable types.

Reply:
Done